php_phongo.c

```diff
@@ -1878,17 +1264,130 @@ static bool php_phongo_apply_wc_options_to_uri(mongoc_uri_t *uri, bson_t *option
+
+	/* Use SSL context options if provided for backwards compatibility */
+	if (php_array_existsc(zdriverOptions, "context")) {
+		php_stream_context *context;
+		zval *zcontext;
+
+		zcontext = php_array_fetchc(zdriverOptions, "context");
+
+		context = php_stream_context_from_zval(zcontext, 1);
+
+		if (!context) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "\"context\" driver option is not a valid Stream-Context resource");
+			return false;
+		}
+
+#if PHP_VERSION_ID >= 70000
+		zoptions = php_array_fetchc_array(&context->options, "ssl");
+#else
+		zoptions = php_array_fetchc_array(context->options, "ssl");
+#endif
+
+		if (!zoptions) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Stream-Context resource does not contain \"ssl\" options array");
+			return false;
+		}
+	}
+
+	/* Check canonical option names first and fall back to SSL context options
+	 * for backwards compatibility. */
+	if (php_array_existsc(zoptions, "allow_invalid_hostname")) {
+		ssl_opt.allow_invalid_hostname = php_array_fetchc_bool(zoptions, "allow_invalid_hostname");
+	}
+
+	if (php_array_existsc(zoptions, "weak_cert_validation")) {
+		ssl_opt.weak_cert_validation = php_array_fetchc_bool(zoptions, "weak_cert_validation");
+	} else if (php_array_existsc(zoptions, "allow_self_signed")) {
+		ssl_opt.weak_cert_validation = php_array_fetchc_bool(zoptions, "allow_self_signed");
+	}
+
+	if (php_array_existsc(zoptions, "pem_file")) {
+		ssl_opt.pem_file = php_array_fetchc_string(zoptions, "pem_file", &plen, &free_pem_file);
+	} else if (php_array_existsc(zoptions, "local_cert")) {
+		ssl_opt.pem_file = php_array_fetchc_string(zoptions, "local_cert", &plen, &free_pem_file);
+	}
+
+	if (php_array_existsc(zoptions, "pem_pwd")) {
+		ssl_opt.pem_pwd = php_array_fetchc_string(zoptions, "pem_pwd", &plen, &free_pem_pwd);
+	} else if (php_array_existsc(zoptions, "passphrase")) {
+		ssl_opt.pem_pwd = php_array_fetchc_string(zoptions, "passphrase", &plen, &free_pem_pwd);
+	}
+
+	if (php_array_existsc(zoptions, "ca_file")) {
+		ssl_opt.ca_file = php_array_fetchc_string(zoptions, "ca_file", &plen, &free_ca_file);
+	} else if (php_array_existsc(zoptions, "cafile")) {
+		ssl_opt.ca_file = php_array_fetchc_string(zoptions, "cafile", &plen, &free_ca_file);
+	}
+
+	if (php_array_existsc(zoptions, "ca_dir")) {
+		ssl_opt.ca_dir = php_array_fetchc_string(zoptions, "ca_dir", &plen, &free_ca_dir);
+	} else if (php_array_existsc(zoptions, "capath")) {
+		ssl_opt.ca_dir = php_array_fetchc_string(zoptions, "capath", &plen, &free_ca_dir);
+	}
+
+	if (php_array_existsc(zoptions, "crl_file")) {
+		ssl_opt.crl_file = php_array_fetchc_string(zoptions, "crl_file", &plen, &free_crl_file);
+	}
+
+	valid_options = true;
+
+#ifdef MONGOC_ENABLE_SSL_OPENSSL
+	if (ssl_opt.ca_dir) {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "\"ca_dir\" or \"capath\" options may only be used with OpenSSL");
+		valid_options = false;
+	}
+#endif
```

Comment:
This logic was actually incorrect. We should error if  `MONGOC_ENABLE_SSL_OPENSSL` was not defined.

Since libmongoc 1.5.0-dev now has LibreSSL support, a more correct approach would be to error based on whether Secure Channel or Secure Transport are being used. I'll address this in https://github.com/mongodb/mongo-php-driver/pull/376.


 Forgot to remove this before merging. Will be addressed in https://github.com/mongodb/mongo-php-driver/pull/376.